src/BloomExe/web/ServerBase.cs

```diff
@@ -116,117 +73,119 @@ private bool StartWithExceptionHandling(out Exception error)
+
 				if (0 == WaitHandle.WaitAny(new[] { _stop, context.AsyncWaitHandle }))
 					return;
 			}
 		}
 
-		/// <summary>
-		/// TODO: Note: doing this at runtime isn't as good as doing it in the installer, because we have no way of
-		/// removing these entries on uninstall (but the installer does).
-		/// </summary>
-		private static void AddUrlAccessControlEntry()
+		private void ContextReady(IAsyncResult ar)
 		{
-			MessageBox.Show(
-				string.Format("We need to do one more thing before Bloom is ready. Bloom needs temporary administrator privileges to set up part of its communication with the embedded web browser.{0}{0}After you click 'OK', you may be asked to authorize this step.", Environment.NewLine),
-				"Almost there!", MessageBoxButtons.OK);
-
-			var startInfo = new System.Diagnostics.ProcessStartInfo();
-			startInfo.UseShellExecute = true;
-			startInfo.Verb = "runas"; //makes it as for elevation to admin rights
-
-			if (Environment.OSVersion.Version.Major == 5 /*win xp*/)
-			{
-				startInfo.FileName = "httpcfg";
-				startInfo.Arguments = string.Format("set urlacl -u {0} /a \"D:(A;;GX;;;WD)\"", PathEndingInSlash);
-			}
-			else
+			try
 			{
-				startInfo.FileName = "netsh";
-				startInfo.Arguments = string.Format("http add urlacl url={0} user=everyone", PathEndingInSlash.TrimEnd('/'));
+				lock (_queue)
+				{
+					_queue.Enqueue(_listener.EndGetContext(ar));
+					_ready.Set();
+				}
 			}
-			System.Diagnostics.Process.Start(startInfo);
+			catch { }
 		}
 
-		private void ListenerCallback(IAsyncResult ar)
+		private void Worker()
 		{
-			if (_isDisposing || _listener == null || !_listener.IsListening)
-				return; //strangely, this callback is fired when we close down the listener
-			string rawurl="unknown";
-			try
-			{
-				HttpListenerContext context = _listener.EndGetContext(ar);
-				rawurl = context.Request.RawUrl;
-				HttpListenerRequest request = context.Request;
-				MakeReply(new RequestInfo(context));
-			}
-			catch(HttpListenerException e)
-			{
-				//http://stackoverflow.com/questions/4801868/c-sharp-problem-with-httplistener
-				Logger.WriteEvent("At ServerBase: ListenerCallback(): HttpListenerException, which may indicate that the caller closed the connection before we could reply. msg=" + e.Message);
-				Logger.WriteEvent("At ServerBase: ListenerCallback(): url=" + rawurl);
-			}
-			catch (Exception error)
+			WaitHandle[] wait = { _ready, _stop };
+			while (0 == WaitHandle.WaitAny(wait))
 			{
-				Logger.WriteEvent("At ServerBase: ListenerCallback(): msg="+ error.Message);
-				Logger.WriteEvent("At ServerBase: ListenerCallback(): url="+rawurl);
-				Logger.WriteEvent("At ServerBase: ListenerCallback(): stack=");
-				Logger.WriteEvent(error.StackTrace);
+				HttpListenerContext context;
+				lock (_queue)
+				{
+					if (_queue.Count > 0)
+						context = _queue.Dequeue();
+					else
+					{
+						_ready.Reset();
+						continue;
```

Comment:
Looking at the documentation of WaitAny, it looks like if _stop is received it will break out of the while loop (https://msdn.microsoft.com/en-us/library/tdykks7z%28v=vs.100%29.aspx).